pkg/kubelet/kubelet_getters.go

```diff
@@ -357,17 +357,46 @@ func (kl *Kubelet) getMountedVolumePathListFromDisk(podUID types.UID) ([]string,
 	return mountedVolumes, nil
 }
 
-// podVolumesSubpathsDirExists returns true if the pod volume-subpaths directory for
-// a given pod exists
-func (kl *Kubelet) podVolumeSubpathsDirExists(podUID types.UID) (bool, error) {
-	podVolDir := kl.getPodVolumeSubpathsDir(podUID)
+// podVolumesSubpathsDirExists returns a list of the volume-subpath paths by reading the
+// subpath directories for the given pod from the disk.
+func (kl *Kubelet) getPodVolumeSubpathListFromDisk(podUID types.UID) ([]string, error) {
+	volumes := []string{}
+	podSubpathsDir := kl.getPodVolumeSubpathsDir(podUID)
 
-	if pathExists, pathErr := mount.PathExists(podVolDir); pathErr != nil {
-		return true, fmt.Errorf("error checking if path %q exists: %v", podVolDir, pathErr)
+	if pathExists, pathErr := mount.PathExists(podSubpathsDir); pathErr != nil {
+		return nil, fmt.Errorf("error checking if path %q exists: %v", podSubpathsDir, pathErr)
 	} else if !pathExists {
-		return false, nil
+		return volumes, nil
 	}
-	return true, nil
+
+	// Explicitly walks /<volume>/<container name>/<subPathIndex>
+	volumePluginDirs, err := ioutil.ReadDir(podSubpathsDir)
+	if err != nil {
+		klog.Errorf("Could not read directory %s: %v", podSubpathsDir, err)
+		return volumes, err
+	}
+	for _, volumePluginDir := range volumePluginDirs {
+		volumePluginName := volumePluginDir.Name()
+		volumePluginPath := filepath.Join(podSubpathsDir, volumePluginName)
+		containerDirs, err := ioutil.ReadDir(volumePluginPath)
+		if err != nil {
+			return volumes, fmt.Errorf("could not read directory %s: %v", volumePluginPath, err)
+		}
+		for _, containerDir := range containerDirs {
+			containerName := containerDir.Name()
+			containerPath := filepath.Join(volumePluginPath, containerName)
+			// Switch to ReadDirNoStat at the subPathIndex level to prevent issues with stat'ing
+			// mount points that may not be responsive
+			subPaths, err := utilpath.ReadDirNoStat(containerPath)
```

Comment:
is that possible that containerPath is empty Dir?

Reply:
Does that matter? The code would clean up an empty directory at containerPath. Or do you mean a K8s emptyDir volume? That shouldn't be possible as we're one level too high for that.

Reply:
will the return volumes contains containerPath if it is an emptyDir? The code only remove any dir that is listed in the volumes?

Reply:
The code will never remove a volume which contains even a single file. Even if the volume is not a mount. It intentionally uses the `rmdir` syscall (and its equivalent on Windows) which fails if there is anything still in there.

Reply:
i got that. The question I have here is say volumePluginDir is an empty dir, containerDirs will be empty list, and the returned volumes will be an empty list too?

Reply:
If volumePluginDir is an empty dir then the loop over containerDirs will not execute as there's no items and thus no volumes will be added to the output.

Reply:
so in this case, volumePluginDir will not be removed. Is this desired behavior or we also want to remove the dir if it is empty.

Reply:
It will be removed by the subsequent RemoveAllOneFilesystem call.

Reply:
I see. but since RemoveAllOneFilesystem can remove all dir/files if it can, the purpose of removing dir here is to reduce log spam? But the log will only appears once if RemoveAllOneFilesystem can clean up dir. The case you are trying to fix is RemoveAllOneFilesystem fails, and it will keep logging the message?

Reply:
No, the Kubelet code currently bails and logs an error if any volume directory still exists. `RemoveAllOneFilesystem` is not actually being called.